sycl/include/sycl/ext/oneapi/experimental/work_group_memory.hpp

```diff
@@ -47,7 +50,14 @@ class __SYCL_SPECIAL_CLASS __SYCL_TYPE(work_group_memory) work_group_memory
       value_type, access::address_space::local_space>::type *;
 
 public:
+// Frontend requires special types to have a default constructor in device
+// compilation mode in order to have a uniform way of initializing an object of
+// special type to then call the __init method on it. This is purely an
+// implementation detail and not part of the spec.
+#ifdef __SYCL_DEVICE_ONLY__
   work_group_memory() = default;
+#endif
```

Comment:
I don't think this is legal.  Doesn't it violate ODR?

This is observable to the user, so isn't just an implementation detail.  Exposing a public default constructor allows a user to call `work_group_memory()` in device code, and also causes `std::is_default_constructible_v<work_group_memory>` to produce different values on host and device (which could break all sorts of things).

One fix would be to make this a protected function instead, and then create a factory function in the `detail::` namespace that creates an uninitialized `work_group_memory` and calls `__init` on it.

Reply:
This compiles fine. Exposing it to the user in device code is a problem, but unfortunately the way our frontend is set-up it requires special types to have a default constructor so that it will initialize them using that constructor in device code and then call `__init` on it. I did explore other avenues but the frontend team, @elizabethandrews , really don't like to have different handlings for different special types. Presumably, this sets a bad precedent for the future where other special types will also have their own hacks like this and the code will be non-uniform and harder to reason about.
Surprisingly, another class which does not have a default constructor as per the spec but seems to add one in device compilation mode is `sycl::stream`.

Reply:
I didn't mean to imply that the code wouldn't compile.  I think it's invalid because it results in multiple definitions of the same class (one for the host, one for the device), because it's not what the specification says, and because it could break user code in surprising ways.

I didn't realize we were already doing this for `sycl::stream`.  I don't think we should be doing that, either.

Since the function call is only required by the front-end, can we just mark these default constructors private, and then ask the front-end to ignore the access specifier and call it anyway?

Tagging @gmlueck for awareness.

Reply:
We could probably use some knowledge input from @intel/dpcpp-cfe-reviewers.
Can we do what @Pennycook suggested and have the Semantic analyzer ignore access modifiers just for the duration of this default initialization for special types?
 This would allow us to mark these default constructors private for classes that do not have them in the spec. This would be something that would apply to all special types, so we would preserve uniformity.

Reply:
> I didn't mean to imply that the code wouldn't compile. I think it's invalid because it results in multiple definitions of the same class (one for the host, one for the device), because it's not what the specification says, and because it could break user code in surprising ways.
> 
> I didn't realize we were already doing this for `sycl::stream`. I don't think we should be doing that, either.
> 
> Since the function call is only required by the front-end, can we just mark these default constructors private, and then ask the front-end to ignore the access specifier and call it anyway?
> 
> Tagging @gmlueck for awareness.

I made the requested changes. Have a look now please.
Let's see if this looks ok to @intel/dpcpp-cfe-reviewers .

Reply:
> I am not sure how open the community will be to upstreaming something like this.

I was under impression that upstream will receive another version of kernel generation that will happen in CodeGen, bypassing semantic analysis for the code that is purely generated by the compiler. Is that not correct?

Reply:
We won't necessarily bypass Semantic Analysis I think. We still generate an `OutlinedFunctionDecl` in the AST similar to CUDA in the new design which is used by CodeGen to emit the kernel. We just don't have a standalone `FunctionDecl` anymore. I can't comment on the specifics since we have not had to construct a clone in the new design yet but I anticipate us having the same problems unless we figure out another way to initialize the clone.

Reply:
> I can't comment on the specifics since we have not had to construct a clone in the new design yet but I anticipate us having the same problems unless we figure out another way to initialize the clone.

Does initialization have to be a constructor, or can we generate arbitrary code here?

My other suggestion was to define a protected constructor, and then add something like `detail::make_uninitialized<work_group_memory>()` as a friend factory function.  Our implementation (including the front-end?) could call this factory function anywhere it needs to default-construct a `work_group_memory` object (or any other special SYCL object).

Reply:
>Does initialization have to be a constructor, or can we generate arbitrary code here?

I think we can generate arbitrary code. I don't think we should generate different code for different special types though. The handling of special types is supposed to be uniform across types. If we can just modify all special types to use the factory function when default constructors are not available, I think that works. @premanandrao @tahonermann thoughts?

Reply:
We don't yet have a design for how to handle SYCL special types and their unique marshaling requirements in what we are sending upstream. We'll be starting to look at that soon though. I have been thinking about this issue for a while and I do think design changes are warranted. I have a few ideas, but no concrete proposal to offer at this point. Regardless, none of that will help with what needs to be done here in the short term.

As for whether this is an ODR issue, I don't think there is a good answer. The C++ standard doesn't offer any guidance since the concept of host vs device is out of scope. SYCL execution can be thought of as multiple programs concurrently running in a shared address space.

My short term suggestion is to accept this for now. There is precedent in the form of `sycl::stream`, so not adding yet another way to work around a design limitation is advantageous. We can then address this as part of the upstreaming effort.

Reply:
> My short term suggestion is to accept this for now. There is precedent in the form of `sycl::stream`, so not adding yet another way to work around a design limitation is advantageous. We can then address this as part of the upstreaming effort.

How quickly do you think we will be able to address this, if it's part of the upstreaming effort?  My concern is that we won't give this issue a very high priority, and in the meantime we'll end up introducing more special types (increasing our chances of somebody being tripped up by this).  I'm not comfortable with the idea of adding new classes that we know are broken unless we also know how long it's going to take to fix them.

Even if the C++ specification isn't clear about whether this is an ODR violation, it's clearly weird and has the potential to break user code.  This could break use of concepts, SFINAE overloads, etc, and the user won't get a good error message.  Consider the below (also on https://godbolt.org/z/M3ze1xx13):
```c++
#include <sycl/sycl.hpp>
#include <type_traits>

static_assert(std::is_default_constructible_v<sycl::stream>);     // asserts in the host pass
static_assert(not std::is_default_constructible_v<sycl::stream>); // asserts in the device pass
```

Reply:
I share @Pennycook's concerns, but I have an additional concern too.  We specifically removed the default constructor for `work_group_memory` to dissuade people from thinking that they could statically construct one of these objects inside device code.  This was the motivation for the `indeterminate` constructor that you are adding in this PR.

By adding the default constructor back, we defeat the purpose of the `indeterminate` constructor.  We *want* applications to get an error if they try to default-construct a `work_group_memory` object.

My preference for this PR is either:

* Figure out the long-term code pattern we want for these "special" types and use that to implement `work_group_memory`.  We can then open an issue to fix the other special types like `stream`, or
* Add some temporary solution like what you had before and open an issue to figure out the right long term solution.

Reply:
> My other suggestion was to define a protected constructor, and then add something like detail::make_uninitialized<work_group_memory>() as a friend factory function. Our implementation (including the front-end?) could call this factory function anywhere it needs to default-construct a work_group_memory object (or any other special SYCL object).

An alternative approach here is to have some static `__init` method which would accept regular `__init` arguments and do both: construction and initialization of an object.

Perhaps it would be easier to emit a call to a static member function attached to the special type and if construction and initialization both happen together within headers, then it would be less code for FE to emit.

Reply:
I'm concerned about this because we might not be able to remove the default constructor until the next major release, when we are allowed to make API breaking changes.  If we need a temporary solution, I'd rather choose one that does not add a public API that we will need to wait before we can remove.

Reply:
> I'm concerned about this because we might not be able to remove the default constructor until the next major release, when we are allowed to make API breaking changes. If we need a temporary solution, I'd rather choose one that does not add a public API that we will need to wait before we can remove.

We are allowed to break API of `experimental` extensions anytime.

Reply:
>>How quickly do you think we will be able to address this, if it's part of the upstreaming effort? 

I don't anticipate us having special types figured out for upstreaming for a couple of months.  If that timeline is too long to keep the default constructor, we can do one of 2 things for now -  

1. Keep the hack this PR implements in the FE to ignore access specifiers. I THINK this is ok but I do not know if this will result in unintended consequences. 
2. Implement the factory function. This would require additional code for the FE to generate the factory function and call it instead of the constructor while initializing the local clone.

The second option is less hacky but I am not opposed to the first option since we may end up removing all of this in the next 2-3 months anyway. I would like the other FE reviewers to confirm they are ok with the first option if we choose to go down that route.

Reply:
It seemed like we had a temporary solution in a previous commit in this PR which did not require exposing a public member function.  Is there a reason we can't go back to that for our temporary solution until we have a permanent one?

Reply:
Sorry, our comments crossed.  I'm not opposed to your option (2) if that seems less hacky.  If we do that, can we put the factory function in the `detail` namespace?

Reply:
I had missed the fact that code changes had already been drafted to work around the default constructor access. I think that is reasonable as a hacky short term solution.

My preference is to spend as little FE engineer time as possible working around this issue since whatever we do for it will be throw away work. I'd rather spend that time designing the better solution that we'll upstream and migrate to later.

Reply:
> Sorry, our comments crossed. I'm not opposed to your option (2) if that seems less hacky. If we do that, can we put the factory function in the `detail` namespace?

I think so? I'm just not sure the effort required for this off the top of my head. As Tom mentioned, it may just not be worth it at this point. We could just go with option 1 adding a comment along the lines of "This is a temporary hack" and focus on designing whatever we upstream.

Reply:
Either (1) or (2) is fine with me.  I'm happy as long as we don't expose public APIs to the user which are not official public APIs.

Reply:
Personally, I don't see (1) as any more of a hack than the use of the `__init` function as a whole. It is the compiler making the decision to use some member function(s) of the type, by instruction of the directives we give it. Luckily it isn't a user-facing feature, so we only have our selves to blame if we mess it up.

Note also that the `__init()` member function can be - and is in most classes - declared private. @lbushi25 - Could we also move that to private while we are at it?

@tahonermann & @elizabethandrews - SYCLSpecialClassDocs in clang/include/clang/Basic/AttrDocs.td does not mention the default ctor requirement, though it has been a requirement for as long as I can remember. Should we amend it and, if we go with (1) here, note the access specifier relaxation?

Reply:
I don't think we need to update the documentation. Comments in the code should suffice since it is an incorrect implementation assumption which we need to change eventually.

Reply:
I agree with @elizabethandrews. We'll need to revisit the docs with whatever design we publish upstream anyway.